mypyc/irbuild/format_str_tokenizer.py

```diff
@@ -30,6 +44,74 @@ def tokenizer_printf_style(format_str: str) -> Tuple[List[str], List[ConversionS
     return literals, specifiers
 
 
+# The empty Context as an argument for parse_format_value().
+# It wouldn't be used since the code has passed the type-checking.
+EMPTY_CONTEXT: Final = Context()
+
+
+def tokenizer_format_call(
+        format_str: str) -> Optional[Tuple[List[str], List[ConversionSpecifier]]]:
+    """Tokenize a str.format() format string.
+
+    The core function parse_format_value() is shared with mypy.
+    However, this function
+
+    Return:
+        A list of string literals and a list of conversion operations.
+        Return None if it cannot parse the string.
+    """
+    # Creates an empty MessageBuilder here.
+    # It wouldn't be used since the code has passed the type-checking.
+    specifiers = parse_format_value(format_str, EMPTY_CONTEXT,
+                                    MessageBuilder(Errors(), {}))
+    if specifiers is None:
+        return None
+
+    literals: List[str] = []
+    last_end = 0
+    for spec in specifiers:
+        # Skip { and }
+        literals.append(format_str[last_end:spec.start_pos - 1])
+        last_end = spec.start_pos + len(spec.whole_seq) + 1
+    literals.append(format_str[last_end:])
+
+    # Deal with escaped {{
+    literals = [x.replace('{{', '{').replace('}}', '}') for x in literals]
+    return literals, specifiers
+
+
+def generate_format_ops(specifiers: List[ConversionSpecifier]) -> Optional[List[FormatOp]]:
+    format_ops = []
+    for spec in specifiers:
+        # TODO: Match specifiers instead of using whole_seq
+        if spec.whole_seq == '%s' or spec.whole_seq == '{:{}}':
```

Comment:
What about `%d`? Will it be supported in a follow-up PR?

It would be better to make this generic over different kinds of formatting instead of special casing percent sign formatting, etc. Do you plan to improve this in a follow-up PR (I assume that this is what the TODO comment is about)?

Reply:
yes